src/Sylius/Bundle/ApiBundle/Validator/Constraints/OrderProductEligibilityValidator.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+final class OrderProductEligibilityValidator extends ConstraintValidator
+{
+    /** @var OrderRepositoryInterface */
+    private $orderRepository;
+
+    public function __construct(OrderRepositoryInterface $orderRepository)
+    {
+        $this->orderRepository = $orderRepository;
+    }
+
+    /**
+     * @throws \InvalidArgumentException
+     */
+    public function validate($command, Constraint $constraint): void
+    {
+        $order = $this->orderRepository->findOneBy(['tokenValue' => $command->orderTokenValue]);
```

Comment:
You could assert earlier if `$command` is a proper value (probably `OrderTokenValueAwareInterface`)